### PR TITLE
API-37794-brd-handle-timeouts 

### PR DESCRIPTION
**Pull request by stiehlrod**
## Summary

- Adds lighthouse timeout error. 
- Catches and raises timeout, and 
- returns an error explaining that an upstream service is timing out. 
- Adds library base so BD can also utilize this error handling.

## Related issue(s)

- [API-37794](https://jira.devops.va.gov/browse/API-37794)

## Testing done

- [X] *New code is covered by unit tests*
- Postman: 526 v2 sync, claims v2 index & show

## What areas of the site does it impact?
	modified:   modules/claims_api/lib/brd/brd.rb
	new file:   modules/claims_api/lib/claims_api/common/exceptions/lighthouse/timeout.rb
	modified:   modules/claims_api/lib/claims_api/v2/error/lighthouse_error_handler.rb
	modified:   modules/claims_api/lib/custom_error.rb
	new file:   modules/claims_api/lib/library_base.rb

## Acceptance criteria

- [X]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [ ]  No error nor warning in the console.
- [ ]  Events are being sent to the appropriate logging solution
- [ ]  Documentation has been updated (link to documentation)
- [x]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ]  Feature/bug has a monitor built into Datadog (if applicable)
- [ ]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [ ]  I added a screenshot of the developed feature

## Requested Feedback

@tycol7 I need to switch my focus, but maybe you can poke holes in this idea/see if it could be used for BD as well as BRD.

### CODE REVIEW

File: modules/claims_api/lib/library_base.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
+# frozen_string_literal: true
+
+require 'claims_api/claim_logger'
+require 'custom_error'
+
+module ClaimsApi
+  class LibraryBase
+    def error_handler(error, async = true) # rubocop:disable Style/OptionalBooleanParameter
+      ClaimsApi::CustomError.new(error, async).build_error
```
Comment (tycol7) [inline modules/claims_api/lib/library_base.rb:9]: In what case might `async` be `false`? As written, we only call `error_handler` on line 25 with one param, meaning `async` will always be `true`. Would it make sense to write something like:

```suggestion
    def error_handler(error)
      ClaimsApi::CustomError.new(error, true).build_error
```

or even more succinctly, replace line 25 with `ClaimsApi::CustomError.new(e, true).build_error` and remove this method?

File: modules/claims_api/lib/library_base.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    end
+
+    def log_outcome_for_claims_api(action, status, response)
+      ClaimsApi::Logger.log(action,
+                            detail: "#{action} #{status}: #{response}")
+    end
+
+    def handle_transaction(service)
+      resp = client.get(service).body[:items]
```
Comment (tycol7) [inline modules/claims_api/lib/library_base.rb:18]: Curious if the intent is to use this method for `post` requests as well. If so, would we need a param to indicate the request type + associated logic? If not, should we rename this method to something like `handle_get_request`?

File: modules/claims_api/lib/library_base.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+      log_outcome_for_claims_api(service, 'success', resp)
```
Comment (tycol7) [inline modules/claims_api/lib/library_base.rb:19]: This line adds new functionality (logging successful requests). Do we want these logs?

**Review by tycol7 — COMMENTED** [review]
> @tycol7 I need to switch my focus, but maybe you can poke holes in this idea/see if it could be used for BD as well as BRD.

Overall, makes sense to abstract timeout handling. Had a few questions on the `LibraryBase` class, but otherwise looks good and could use the custom error handling for BD.

File: modules/claims_api/lib/library_base.rb (hunk 3 above)
Comment (stiehlrod) [reply modules/claims_api/lib/library_base.rb:19]: @tycol7 Sometimes it is nice to see which calls where successful, especially when the service gets called multiple times for one endpoint. Debatable, for sure.

Comment (FonzMP) [reply modules/claims_api/lib/library_base.rb:10]: I agree here that this extra step is probably not needed. additionally, some of these responses from BRD can be massive as well, not sure we want to log all of that on every request - /disabilities in staging has over 4,000 lines. On top of that, we make multiple calls to BRD at the current time, so you're talking 3'ish times the expected volume of 526 submissions (roughly 4,000+ a day)... that's 12,000+ logs of their response in it's current state.

As straightforward as possible, I'd suggest in rescuing a `Net::HTTPGatewayTimeout` in the service as mentioned in the ticket, and then raise a `Lighthouse::Timeout`(assuming that template is aligning with [LH standards](https://department-of-veterans-affairs.github.io/va-api-standards/errors/)). 
Do we know how much that differs from how the `::Common::Exceptions::GatewayTimeout`? Think I would still lean a new one as it's best if we can control / change the error - plus we'll be able to easily extract just the LH errors when we move to LHDI.

File: modules/claims_api/lib/claims_api/common/exceptions/lighthouse/timeout.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+          end
+
+          def status_code
+            504
+          end
+        end
+      end
+    end
+  end
+end
```
Comment (rockwellwindsor-va) [inline modules/claims_api/lib/claims_api/common/exceptions/lighthouse/timeout.rb:30]: @stiehlrod  Would it be valuable to add a test in the `lighthouse_error_handler_spec` file for this error?

File: modules/claims_api/lib/claims_api/common/exceptions/lighthouse/timeout.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
+module ClaimsApi
+  module Common
+    module Exceptions
+      module Lighthouse
+        class Timeout < StandardError
+          def initialize(errors)
+            @errors = [errors].flatten
```
Comment (FonzMP) [inline modules/claims_api/lib/claims_api/common/exceptions/lighthouse/timeout.rb:9]: we do nothing with this currently - is this just providing flexibility for future use cases / expansion?

File: modules/claims_api/lib/brd/brd.rb [hunk 6]
```diff
@@ -22,14 +29,29 @@ def countries
     # as {id: <number> and description: <string>}
     def intake_sites
       client.get('intake-sites').body[:items]
+    rescue => e
+      detail = e.respond_to?(:original_body) ? e.original_body : e
+      log_outcome_for_claims_api(service, 'error', detail)
+
+      error_handler(e)
```
Comment (FonzMP) [inline modules/claims_api/lib/brd/brd.rb:36]: Nitpick: a tidy up ticket could be to wrap these in a method and simplify the rescues a bit.

File: modules/claims_api/lib/custom_error.rb [hunk 7]
```diff
@@ -15,6 +16,12 @@ def build_error
       case @error
       when Faraday::ParsingError
         raise_backend_exception
+      when ::Common::Exceptions::GatewayTimeout,
+        Timeout::Error,
+        Faraday::TimeoutError,
+        Breakers::OutageException
```
Comment (FonzMP) [inline modules/claims_api/lib/custom_error.rb:22]: we'll still need to handle the Net::HTTPGatewayTimeout as mentioned in the ticket.

File: modules/claims_api/lib/claims_api/common/exceptions/lighthouse/timeout.rb (hunk 5 above)
Comment (stiehlrod) [reply modules/claims_api/lib/claims_api/common/exceptions/lighthouse/timeout.rb:9]: @FonzMP We use it here in the [lighthouse_error_handler](https://github.com/department-of-veterans-affairs/vets-api/blob/API-37794-brd-handle-timeouts/modules/claims_api/lib/claims_api/v2/error/lighthouse_error_handler.rb#L29)

Comment (FonzMP) [reply modules/claims_api/lib/claims_api/common/exceptions/lighthouse/timeout.rb:9]: I'm meaning line 9. 
`@errors = [errors].flatten`

we don't use @errors again, right?

File: modules/claims_api/spec/lib/claims_api/v2/error/lighthouse_error_handler_spec.rb [hunk 8]
```diff
@@ -47,6 +47,11 @@ def raise_invalid_token
     def raise_expired_token_signature
       raise Common::Exceptions::TokenValidationError.new(detail: 'Signature has expired')
     end
+
+    def raise_timeout
+      error = [{ status_code: 499 }]
```
Comment (FonzMP) [inline modules/claims_api/spec/lib/claims_api/v2/error/lighthouse_error_handler_spec.rb:52]: curious - why did we add this line?

File: modules/claims_api/lib/brd/brd.rb [hunk 9]
```diff
@@ -54,5 +64,12 @@ def client
         f.adapter Faraday.default_adapter
       end
     end
+
+    def rescue_brd(e, service)
+      detail = e.respond_to?(:original_body) ? e.original_body : e
+      log_outcome_for_claims_api(service, 'error', detail)
```
Comment (FonzMP) [inline modules/claims_api/lib/brd/brd.rb:70]: future update: I'd say we may want to elaborate it's the BRD call here in the service for simplicities sake, but that can be a later ticket. Just thinking `intake-sites` may make sense to us, but something like `BRD-intake-sites` could be a good elaboration later and simplify where this is coming from. Granted DataDog does a good job at calling out where in the code base it's coming from. Maybe it's nitpick more than future update.

File: modules/claims_api/lib/custom_error.rb [hunk 10]
```diff
@@ -34,6 +42,11 @@ def raise_backend_exception
       raise ::ClaimsApi::Common::Exceptions::Lighthouse::BackendServiceException, error_details
     end
 
+    def raise_timeout_exception
+      error_details = get_error_info if @original_body.present?
```
Comment (FonzMP) [inline modules/claims_api/lib/custom_error.rb:46]: unused but doesn't hurt anything.

**Review by FonzMP — APPROVED** [review]
:shipit:

**Pull request merged**